### PR TITLE
htmlCapture Nightwatch command

### PR DESCRIPTION
**Pull request by chrisjcook**
Status: **Opened for visibility**

Reviewers: @mobify-derrick @twangtina @ellenmobify @jgermyn @gsaray

To Test:

1.  Create a branch off of any Adaptive project and change the "nightwatch-commands" line in the pacakge.json file to:
> "nightwatch-commands": "git://github.com/mobify/nightwatch-commands.git#nightwatch-commands-htmlcapture"

2.  Run "npm install" to get the new dependencies

3.  In your "nightwatch.json" file, add a new "test_settings" entry for "desktop". This should be the same as your default, but with no "--user-agent" or "--window-size" entries.

4.  You can now use the htmlCapture command in any Nightwatch test. Here is an absolute bare-bones test using the command:
```javascript
module.exports = {
    'home.html Fixture': function(browser) {
        browser
        	.get("http://www.carnival.com")
            .htmlCapture('home.html')
        .end();
    }
};
```

The command will grab and save the HTML of whatever page Nightwatch is currently testing against. It will save the file as whatever you specify in '.htmlCapture("namehere.html")' to the "tests/fixtures/" folder for future integration tests. So for pages that need a workflow to get to, either put the htmlCapture command after the navigation workflow or create a separate, dedicated test with the same workflow (recommended).

Since you'll be creating a separate test to pull HTML fixtures from each page (and therefore each Nightwatch file), it's going to be more manageable to have all of these tests in one place. Also, since we want to be capturing desktop HTML source, we need to use htmlCapture independently of any mobile tests. Currently, I've created a new folder in tests/system/ called "htmlfixtures", which contains two files, "static.js" and "workflow.js". The "static.js" contains pages that are accessible by previewing or navigating to directly, whereas "workflow.js" contains pages like Checkout and Account that need a workflow to get to. Here's what that file structure looks like:

htmlfixtures-folder

*Remember:* Make sure to include your "desktop" environment setting in your nightwatch command. IE:

`grunt nightwatch -test "tests/system/htmlcapture/static.js" -e "desktop"`

### CODE REVIEW

File: commands/htmlCapture.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        if ( e.code != 'EEXIST' ) throw e;
+    }
+}
+
+exports.command = function(fileName, callback) {
+    if (fileName && typeof fileName !== 'string') {
+        throw new Error('htmlCapture expects first parameter to be string; ' + typeof (fileName) + ' given');
+    }
+
+    var htmlCheck = /^\w+\.html$/;
```
Comment (ellenmobify) [inline commands/htmlCapture.js]: Saved fixtures will always be .html files. We can probably relax this regex and make it optional for the user to specify ".html" in the file name.

File: commands/htmlCapture.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    if (!htmlCheck.test(fileName)){
+        throw new Error('htmlCapture expects first parameter to be camelCased alphanumeric string ending with ".html"');
+    }
+
+    var client = this;
+    var filePath = "tests/fixtures/";
+    var fileLocation = filePath + fileName;
```
Comment (chrisjcook) [inline commands/htmlCapture.js]: This should be tested/modified to eliminate the possibility of the folder being created somewhere other than intended.

**Comment by ellenmobify** [comment]
There's an issue where `.source` grabs the page source before the page is completely loaded - specifically for dynamic content. Possible ways to get around this would be to put the capture command after checking for the existence of dynamically loaded content, hardcoding pauses, checking for active ajax requests.

File: commands/htmlCapture.js (hunk 1 above)
Comment (chrisjcook) [inline commands/htmlCapture.js]: I was thinking about this, and I feel like having the ".html" be necessary makes the code more readable. ie: 
```javascript
browser
    .htmlCapture("home")
    .end();
```
doesn't imply functionality as well as
```javascript
    .htmlCapture("home.html")
```
Your thoughts?

**Comment by ellenmobify** [comment]
:+1: Works as expected with the desktop environment.

**Pull request merged**